File: BreathTrainer.cs
# BreathTrainer.cs
/**
 * 🌬️ Breath Trainer – 4-7-8 Relaxation Technique (C# Edition)
 * Features: guided breathing, progress bars, stats, custom timing, sound
 * Requires: .NET 6.0+
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;

class BreathTrainer
{
    // ─── Colors ────────────────────────────────────────────────────────────

    private static readonly string Reset = "\u001B[0m";
    private static readonly string Bright = "\u001B[1m";
    private static readonly string Dim = "\u001B[2m";
    private static readonly string Red = "\u001B[31m";
    private static readonly string Green = "\u001B[32m";
    private static readonly string Yellow = "\u001B[33m";
    private static readonly string Blue = "\u001B[34m";
    private static readonly string Magenta = "\u001B[35m";
    private static readonly string Cyan = "\u001B[36m";

    private static string C(string text, string color) => color + text + Reset;

    // ─── Sound ─────────────────────────────────────────────────────────────

    private static void Beep()
    {
        try
        {
            Console.Beep(800, 200);
        }
        catch
        {
            Console.Write('\u0007');
        }
    }

    // ─── Data Model ──────────────────────────────────────────────────────

    public class Timing
    {
        [JsonPropertyName("inhale")]
        public int Inhale { get; set; } = 4;
        [JsonPropertyName("hold")]
        public int Hold { get; set; } = 7;
        [JsonPropertyName("exhale")]
        public int Exhale { get; set; } = 8;
    }

    public class Session
    {
        [JsonPropertyName("date")]
        public string Date { get; set; } = "";
        [JsonPropertyName("cycles")]
        public int Cycles { get; set; }
        [JsonPropertyName("durationSeconds")]
        public int DurationSeconds { get; set; }
    }

    public class StatsData
    {
        [JsonPropertyName("totalSessions")]
        public int TotalSessions { get; set; }
        [JsonPropertyName("totalCycles")]
        public int TotalCycles { get; set; }
        [JsonPropertyName("totalTimeSeconds")]
        public int TotalTimeSeconds { get; set; }
        [JsonPropertyName("sessions")]
        public List<Session> Sessions { get; set; } = new();
        [JsonPropertyName("customTiming")]
        public Timing CustomTiming { get; set; } = new();
    }

    // ─── Config ────────────────────────────────────────────────────────────

    private static readonly string DataDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
        ".breath_trainer"
    );
    private static readonly string DataFile = Path.Combine(DataDir, "stats.json");

    // ─── Breath Trainer ──────────────────────────────────────────────────

    private readonly StatsData stats = new();
    private int inhale, hold, exhale;
    private bool soundOn = true;

    public BreathTrainer()
    {
        Directory.CreateDirectory(DataDir);
        Load();
        inhale = stats.CustomTiming.Inhale;
        hold = stats.CustomTiming.Hold;
        exhale = stats.CustomTiming.Exhale;
    }

    private void Load()
    {
        if (!File.Exists(DataFile)) return;
        try
        {
            string json = File.ReadAllText(DataFile);
            var loaded = JsonSerializer.Deserialize<StatsData>(json);
            if (loaded != null)
            {
                stats.TotalSessions = loaded.TotalSessions;
                stats.TotalCycles = loaded.TotalCycles;
                stats.TotalTimeSeconds = loaded.TotalTimeSeconds;
                stats.Sessions = loaded.Sessions ?? new List<Session>();
                stats.CustomTiming = loaded.CustomTiming ?? new Timing();
            }
        }
        catch { /* ignore */ }
    }

    private void Save()
    {
        string json = JsonSerializer.Serialize(stats, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(DataFile, json);
    }

    private void SleepMs(int ms)
    {
        Thread.Sleep(ms);
    }

    private void ShowStage(string stage, int duration, string color, string emoji)
    {
        int steps = duration * 2;
        for (int i = 0; i <= steps; i++)
        {
            double pct = (double)i / steps * 100;
            int filled = (int)((double)i / steps * 30);
            string bar = new string('█', filled) + new string('░', 30 - filled);
            Console.Write($"\r  {emoji} {C(stage, color)} {C(bar, color)} {pct,3:F0}%");
            SleepMs(500);
        }
        Console.WriteLine();
    }

    private double BreathingCycle()
    {
        var start = DateTime.Now;
        if (soundOn) Beep();
        ShowStage("Breathe In...", inhale, Blue, "🌬️");
        if (soundOn) Beep();
        ShowStage("Hold...", hold, Yellow, "⏸️");
        if (soundOn) Beep();
        ShowStage("Exhale...", exhale, Green, "🌊");
        return (DateTime.Now - start).TotalSeconds;
    }

    private void RunSession(int cycles)
    {
        Console.WriteLine(C($"\n🌬️ 4-7-8 Breathing Session – {cycles} cycles", Bright + Cyan));
        if (soundOn) Beep();
        Console.WriteLine(C("Get ready...", Dim));
        SleepMs(1000);

        double totalDuration = 0;
        for (int i = 0; i < cycles; i++)
        {
            Console.WriteLine(C($"\nCycle {i+1}/{cycles}", Magenta));
            totalDuration += BreathingCycle();
        }

        if (soundOn) { Beep(); Beep(); Beep(); }
        Console.WriteLine(C("\n✨ Session complete! ✨", Green));
        Console.WriteLine($"  Cycles: {cycles}");
        Console.WriteLine($"  Duration: {totalDuration:F1} seconds");

        stats.TotalSessions++;
        stats.TotalCycles += cycles;
        stats.TotalTimeSeconds += (int)totalDuration;
        stats.Sessions.Add(new Session
        {
            Date = DateTime.Now.ToString("o"),
            Cycles = cycles,
            DurationSeconds = (int)totalDuration
        });
        if (stats.Sessions.Count > 50) stats.Sessions.RemoveAt(0);
        Save();
    }

    private void ShowStats()
    {
        Console.WriteLine("\n📊 STATISTICS");
        Console.WriteLine(C(new string('─', 30), Dim));
        Console.WriteLine($"  Total Sessions: {stats.TotalSessions}");
        Console.WriteLine($"  Total Cycles:   {stats.TotalCycles}");
        int mins = stats.TotalTimeSeconds / 60;
        int secs = stats.TotalTimeSeconds % 60;
        Console.WriteLine($"  Total Time:     {mins}m {secs}s");
        if (stats.Sessions.Count > 0)
        {
            var last = stats.Sessions.Last();
            Console.WriteLine($"  Last Session:   {last.Cycles} cycles ({last.DurationSeconds}s)");
            Console.WriteLine("\n📅 Recent Sessions:");
            var recent = stats.Sessions.Skip(Math.Max(0, stats.Sessions.Count - 5));
            foreach (var s in recent)
            {
                string d = s.Date[..16].Replace('T', ' ');
                Console.WriteLine($"  {d}  {s.Cycles} cycles  {s.DurationSeconds}s");
            }
        }
    }

    private void SetTiming()
    {
        int inh = AskInt($"Inhale duration (default {inhale}): ");
        int h = AskInt($"Hold duration (default {hold}): ");
        int exh = AskInt($"Exhale duration (default {exhale}): ");
        if (inh > 0) inhale = inh;
        if (h > 0) hold = h;
        if (exh > 0) exhale = exh;
        stats.CustomTiming = new Timing { Inhale = inhale, Hold = hold, Exhale = exhale };
        Save();
        Console.WriteLine(C($"✅ Timing set: {inhale}s inhale, {hold}s hold, {exhale}s exhale", Green));
    }

    private void ToggleSound()
    {
        soundOn = !soundOn;
        Console.WriteLine(C($"🔊 Sound {(soundOn ? "on" : "off")}", Cyan));
    }

    private void ResetStats()
    {
        Console.Write("⚠️  Reset all statistics? (yes/no): ");
        string ans = Console.ReadLine()?.Trim() ?? "";
        if (ans.Equals("yes", StringComparison.OrdinalIgnoreCase))
        {
            stats.TotalSessions = 0;
            stats.TotalCycles = 0;
            stats.TotalTimeSeconds = 0;
            stats.Sessions.Clear();
            Save();
            Console.WriteLine(C("🗑️  Statistics reset.", Yellow));
        }
    }

    // ─── Menu ──────────────────────────────────────────────────────────────

    private string Ask(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private int AskInt(string prompt)
    {
        while (true)
        {
            string ans = Ask(prompt);
            if (string.IsNullOrEmpty(ans)) return 0;
            if (int.TryParse(ans, out int val)) return val;
            Console.WriteLine(C("❌ Please enter a number.", Red));
        }
    }

    private void ShowMenu()
    {
        Console.WriteLine("\n" + C(new string('═', 50), Cyan));
        Console.WriteLine(C("🌬️ BREATH TRAINER", Bright + Cyan));
        Console.WriteLine(C(new string('═', 50), Cyan));
        Console.WriteLine($"  4-7-8: Inhale {inhale}s, Hold {hold}s, Exhale {exhale}s");
        Console.WriteLine($"  Sound: {(soundOn ? "🔊 On" : "🔇 Off")}");
        Console.WriteLine(C(new string('═', 50), Cyan));
        Console.WriteLine("  1. 🌬️ Start Session (4 cycles)");
        Console.WriteLine("  2. 🌬️ Custom Session (choose cycles)");
        Console.WriteLine("  3. 📊 Statistics");
        Console.WriteLine("  4. ⏱️ Set Timings");
        Console.WriteLine("  5. 🔇 Toggle Sound");
        Console.WriteLine("  6. 🗑️ Reset Statistics");
        Console.WriteLine("  0. 🚪 Exit");
        Console.WriteLine(C(new string('═', 50), Cyan));
    }

    public void Run()
    {
        Console.Clear();
        Console.WriteLine(C("\n🌬️ Breath Trainer – 4-7-8 Relaxation", Bright + Cyan));
        Console.WriteLine(C("Breathe in peace, hold calm, exhale stress.", Dim));

        while (true)
        {
            ShowMenu();
            string choice = Ask("Your choice: ");
            switch (choice)
            {
                case "1":
                    RunSession(4);
                    break;
                case "2":
                    int cycles = AskInt("Number of cycles (default 4): ");
                    if (cycles <= 0) cycles = 4;
                    RunSession(cycles);
                    break;
                case "3":
                    ShowStats();
                    break;
                case "4":
                    SetTiming();
                    break;
                case "5":
                    ToggleSound();
                    break;
                case "6":
                    ResetStats();
                    break;
                case "0":
                    Console.WriteLine(C("👋 Breathe deeply! Goodbye!", Cyan));
                    return;
                default:
                    Console.WriteLine(C("❌ Invalid choice.", Red));
                    break;
            }
            if (choice != "0")
            {
                Console.Write("\nPress Enter to continue...");
                Console.ReadLine();
            }
        }
    }

    public static void Main()
    {
        try
        {
            new BreathTrainer().Run();
        }
        catch (Exception ex)
        {
            Console.WriteLine(C($"❌ Unexpected error: {ex.Message}", Red));
            Environment.Exit(1);
        }
    }
}

File: BreathTrainer.java
# BreathTrainer.java
/**
 * 🌬️ Breath Trainer – 4-7-8 Relaxation Technique (Java Edition)
 * Features: guided breathing, progress bars, stats, custom timing, sound
 * Requires: Java 17+
 */

import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import java.util.concurrent.*;

public class BreathTrainer {
    // ─── Colors ────────────────────────────────────────────────────────────

    private static final String RESET = "\u001B[0m";
    private static final String BRIGHT = "\u001B[1m";
    private static final String DIM = "\u001B[2m";
    private static final String RED = "\u001B[31m";
    private static final String GREEN = "\u001B[32m";
    private static final String YELLOW = "\u001B[33m";
    private static final String BLUE = "\u001B[34m";
    private static final String MAGENTA = "\u001B[35m";
    private static final String CYAN = "\u001B[36m";

    private static String c(String text, String color) { return color + text + RESET; }

    // ─── Sound ─────────────────────────────────────────────────────────────

    private static void beep() {
        try {
            java.awt.Toolkit.getDefaultToolkit().beep();
        } catch (Exception e) {
            // Fallback: terminal bell
            System.out.print('\u0007');
            System.out.flush();
        }
    }

    // ─── Data Model ──────────────────────────────────────────────────────

    private static class Timing {
        int inhale, hold, exhale;
        Timing(int inhale, int hold, int exhale) {
            this.inhale = inhale;
            this.hold = hold;
            this.exhale = exhale;
        }
    }

    private static class Session {
        String date;
        int cycles;
        int durationSeconds;
        Session(String date, int cycles, int durationSeconds) {
            this.date = date;
            this.cycles = cycles;
            this.durationSeconds = durationSeconds;
        }
    }

    private static class Stats {
        int totalSessions = 0;
        int totalCycles = 0;
        int totalTimeSeconds = 0;
        List<Session> sessions = new ArrayList<>();
        Timing timing = new Timing(4, 7, 8);
    }

    // ─── Config ────────────────────────────────────────────────────────────

    private static final String DATA_DIR = System.getProperty("user.home") + "/.breath_trainer";
    private static final String DATA_FILE = DATA_DIR + "/stats.json";

    // ─── Breath Trainer ──────────────────────────────────────────────────

    private final Scanner scanner;
    private Stats stats;
    private int inhale, hold, exhale;
    private boolean soundOn = true;

    public BreathTrainer() throws IOException {
        scanner = new Scanner(System.in);
        Files.createDirectories(Paths.get(DATA_DIR));
        stats = new Stats();
        load();
        inhale = stats.timing.inhale;
        hold = stats.timing.hold;
        exhale = stats.timing.exhale;
    }

    private void load() {
        Path path = Paths.get(DATA_FILE);
        if (!Files.exists(path)) return;
        try {
            String json = Files.readString(path);
            // Simple manual parse
            stats.totalSessions = extractInt(json, "totalSessions");
            stats.totalCycles = extractInt(json, "totalCycles");
            stats.totalTimeSeconds = extractInt(json, "totalTimeSeconds");
            int i = extractInt(json, "inhale");
            if (i > 0) stats.timing.inhale = i;
            int h = extractInt(json, "hold");
            if (h > 0) stats.timing.hold = h;
            int e = extractInt(json, "exhale");
            if (e > 0) stats.timing.exhale = e;
        } catch (Exception e) { /* ignore */ }
    }

    private int extractInt(String json, String key) {
        String pattern = "\"" + key + "\"\\s*:\\s*(\\d+)";
        var m = java.util.regex.Pattern.compile(pattern).matcher(json);
        return m.find() ? Integer.parseInt(m.group(1)) : 0;
    }

    private void save() {
        try {
            StringBuilder sb = new StringBuilder();
            sb.append("{\n");
            sb.append("  \"totalSessions\": ").append(stats.totalSessions).append(",\n");
            sb.append("  \"totalCycles\": ").append(stats.totalCycles).append(",\n");
            sb.append("  \"totalTimeSeconds\": ").append(stats.totalTimeSeconds).append(",\n");
            sb.append("  \"sessions\": [\n");
            for (int i = 0; i < stats.sessions.size(); i++) {
                Session s = stats.sessions.get(i);
                sb.append("    { \"date\": \"").append(escapeJson(s.date)).append("\", \"cycles\": ").append(s.cycles).append(", \"durationSeconds\": ").append(s.durationSeconds).append(" }");
                if (i < stats.sessions.size() - 1) sb.append(",");
                sb.append("\n");
            }
            sb.append("  ],\n");
            sb.append("  \"customTiming\": { \"inhale\": ").append(stats.timing.inhale).append(", \"hold\": ").append(stats.timing.hold).append(", \"exhale\": ").append(stats.timing.exhale).append(" }\n");
            sb.append("}");
            Files.writeString(Paths.get(DATA_FILE), sb.toString());
        } catch (IOException e) { e.printStackTrace(); }
    }

    private String escapeJson(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private void sleepMs(int ms) {
        try { Thread.sleep(ms); } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
    }

    private void showStage(String stage, int duration, String color, String emoji) {
        int steps = duration * 2;
        for (int i = 0; i <= steps; i++) {
            double pct = (double) i / steps * 100;
            int filled = (int) ((double) i / steps * 30);
            String bar = "█".repeat(filled) + "░".repeat(30 - filled);
            System.out.printf("\r  %s %s %s %3.0f%%", emoji, c(stage, color), c(bar, color), pct);
            System.out.flush();
            sleepMs(500);
        }
        System.out.println();
    }

    private double breathingCycle() {
        long start = System.currentTimeMillis();
        if (soundOn) beep();
        showStage("Breathe In...", inhale, BLUE, "🌬️");
        if (soundOn) beep();
        showStage("Hold...", hold, YELLOW, "⏸️");
        if (soundOn) beep();
        showStage("Exhale...", exhale, GREEN, "🌊");
        return (System.currentTimeMillis() - start) / 1000.0;
    }

    private void runSession(int cycles) {
        System.out.println(c("\n🌬️ 4-7-8 Breathing Session – " + cycles + " cycles", BRIGHT + CYAN));
        if (soundOn) beep();
        System.out.println(c("Get ready...", DIM));
        sleepMs(1000);

        double totalDuration = 0;
        for (int i = 0; i < cycles; i++) {
            System.out.println(c("\nCycle " + (i+1) + "/" + cycles, MAGENTA));
            totalDuration += breathingCycle();
        }

        if (soundOn) { beep(); beep(); beep(); }
        System.out.println(c("\n✨ Session complete! ✨", GREEN));
        System.out.println("  Cycles: " + cycles);
        System.out.printf("  Duration: %.1f seconds\n", totalDuration);

        stats.totalSessions++;
        stats.totalCycles += cycles;
        stats.totalTimeSeconds += (int) totalDuration;
        stats.sessions.add(new Session(
            Instant.now().toString(),
            cycles,
            (int) totalDuration
        ));
        if (stats.sessions.size() > 50) stats.sessions.remove(0);
        save();
    }

    private void showStats() {
        System.out.println("\n📊 STATISTICS");
        System.out.println(c("─".repeat(30), DIM));
        System.out.println("  Total Sessions: " + stats.totalSessions);
        System.out.println("  Total Cycles:   " + stats.totalCycles);
        int mins = stats.totalTimeSeconds / 60;
        int secs = stats.totalTimeSeconds % 60;
        System.out.println("  Total Time:     " + mins + "m " + secs + "s");
        if (!stats.sessions.isEmpty()) {
            Session last = stats.sessions.get(stats.sessions.size() - 1);
            System.out.println("  Last Session:   " + last.cycles + " cycles (" + last.durationSeconds + "s)");
            System.out.println("\n📅 Recent Sessions:");
            int start = Math.max(0, stats.sessions.size() - 5);
            for (int i = start; i < stats.sessions.size(); i++) {
                Session s = stats.sessions.get(i);
                String d = s.date.substring(0, 16).replace('T', ' ');
                System.out.println("  " + d + "  " + s.cycles + " cycles  " + s.durationSeconds + "s");
            }
        }
    }

    private void setTiming() {
        int inh = askInt("Inhale duration (default " + inhale + "): ");
        int h = askInt("Hold duration (default " + hold + "): ");
        int exh = askInt("Exhale duration (default " + exhale + "): ");
        if (inh > 0) inhale = inh;
        if (h > 0) hold = h;
        if (exh > 0) exhale = exh;
        stats.timing = new Timing(inhale, hold, exhale);
        save();
        System.out.println(c("✅ Timing set: " + inhale + "s inhale, " + hold + "s hold, " + exhale + "s exhale", GREEN));
    }

    private void toggleSound() {
        soundOn = !soundOn;
        System.out.println(c("🔊 Sound " + (soundOn ? "on" : "off"), CYAN));
    }

    private void resetStats() {
        System.out.print("⚠️  Reset all statistics? (yes/no): ");
        String ans = scanner.nextLine().trim();
        if (ans.equalsIgnoreCase("yes")) {
            stats.totalSessions = 0;
            stats.totalCycles = 0;
            stats.totalTimeSeconds = 0;
            stats.sessions.clear();
            save();
            System.out.println(c("🗑️  Statistics reset.", YELLOW));
        }
    }

    // ─── Menu ──────────────────────────────────────────────────────────────

    private String ask(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private int askInt(String prompt) {
        while (true) {
            try {
                String ans = ask(prompt);
                if (ans.isEmpty()) return 0;
                return Integer.parseInt(ans);
            } catch (NumberFormatException e) {
                System.out.println(c("❌ Please enter a number.", RED));
            }
        }
    }

    private void showMenu() {
        System.out.println("\n" + c("═".repeat(50), CYAN));
        System.out.println(c("🌬️ BREATH TRAINER", BRIGHT + CYAN));
        System.out.println(c("═".repeat(50), CYAN));
        System.out.println("  4-7-8: Inhale " + inhale + "s, Hold " + hold + "s, Exhale " + exhale + "s");
        System.out.println("  Sound: " + (soundOn ? "🔊 On" : "🔇 Off"));
        System.out.println(c("═".repeat(50), CYAN));
        System.out.println("  1. 🌬️ Start Session (4 cycles)");
        System.out.println("  2. 🌬️ Custom Session (choose cycles)");
        System.out.println("  3. 📊 Statistics");
        System.out.println("  4. ⏱️ Set Timings");
        System.out.println("  5. 🔇 Toggle Sound");
        System.out.println("  6. 🗑️ Reset Statistics");
        System.out.println("  0. 🚪 Exit");
        System.out.println(c("═".repeat(50), CYAN));
    }

    public void run() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
        System.out.println(c("\n🌬️ Breath Trainer – 4-7-8 Relaxation", BRIGHT + CYAN));
        System.out.println(c("Breathe in peace, hold calm, exhale stress.", DIM));

        while (true) {
            showMenu();
            String choice = ask("Your choice: ");
            switch (choice) {
                case "1": runSession(4); break;
                case "2": {
                    int cycles = askInt("Number of cycles (default 4): ");
                    if (cycles <= 0) cycles = 4;
                    runSession(cycles);
                    break;
                }
                case "3": showStats(); break;
                case "4": setTiming(); break;
                case "5": toggleSound(); break;
                case "6": resetStats(); break;
                case "0":
                    System.out.println(c("👋 Breathe deeply! Goodbye!", CYAN));
                    return;
                default:
                    System.out.println(c("❌ Invalid choice.", RED));
            }
            if (!choice.equals("0")) {
                System.out.print("\nPress Enter to continue...");
                scanner.nextLine();
            }
        }
    }

    public static void main(String[] args) {
        try {
            new BreathTrainer().run();
        } catch (Exception e) {
            System.err.println(c("❌ Unexpected error: " + e.getMessage(), RED));
            e.printStackTrace();
            System.exit(1);
        }
    }
}
